---
title: "kidney-endothelial-ligand-receptor"
author: "yincy"
date: "2/23/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Data load and Preprocessing  
## load packages  
```{r, message=FALSE}
library(Seurat)
library(tidyverse)
library(nichenetr)
library(VennDiagram)
library(RColorBrewer)
library(edgeR)
library(factoextra)
library(limma)
library(igraph)
```


## load ligand-receptor model  
```{r}
ligand_target_matrix <- readRDS(file = "/home/yincy/git/Data/NicheNet/data/ligand_target_matrix.rds")
ligand_target_matrix %>% .[1:5, 1:5]
# ligand in columns, receptor in rows
```


## convert human ligand target model to mouse
```{r}
colnames(ligand_target_matrix) <- ligand_target_matrix %>% 
    colnames() %>% 
    convert_human_to_mouse_symbols()

rownames(ligand_target_matrix) <- ligand_target_matrix %>% 
    rownames() %>% 
    convert_human_to_mouse_symbols()

# remove NAs
ligand_target_matrix <- ligand_target_matrix[!is.na(rownames(ligand_target_matrix)), !is.na(colnames(ligand_target_matrix))]
ligand_target_matrix %>% dim()
```


## process expression data  
```{r Single-cell data from GSE107585}
cluster_id <- read.table(file = "/home/yincy/git/Data/kidney-single-cell/GSE107585_Mouse_kidney_single_cell_datamatrix.txt.gz", nrow = 2) %>% 
    .[1, , drop = T] %>% as.numeric()

reads_table <- read.table(file = "/home/yincy/git/Data/kidney-single-cell/GSE107585_Mouse_kidney_single_cell_datamatrix.txt.gz", 
                          header = FALSE,
                          skip = 2,
                          row.names = 1)

reads_table_pt <- reads_table[, cluster_id == 3]
reads_table_dct <- reads_table[, cluster_id == 5]

reads_table_pt %>% .[1:10, 1:5]
reads_table_dct %>% .[1:5, 1:5]
```


```{r define expressed genes in pt and dct}
# expressed genes in proximal tubule
expressed_genes_pt_id <- apply(reads_table_pt, 1, function(x){mean(x > 0) > 0.05})
table(expressed_genes_pt_id)
expressed_genes_pt <- reads_table_pt[expressed_genes_pt_id, ] %>% rownames()

# expressed genes in distal convoluted tubule
expressed_genes_dct_id <- apply(reads_table_dct, 1, function(x){mean(x > 0) > 0.05})
table(expressed_genes_dct_id)
expressed_genes_dct <- reads_table_dct[expressed_genes_dct_id, ] %>% rownames()
```

```{r}
venn.diagram(x = list(PT = expressed_genes_pt, DCT = expressed_genes_dct), 
             filename = "/home/yincy/git/R-codes/sslab/macrophage/PT-DCT-Venn.tiff", 
             main = "expressed genes in PT and DCT", 
             fill = brewer.pal(n = 3, name = "Paired")[c(1, 3)], 
             alpha = 0.5)
```


```{r Bulk-seq macrophage}
reads_table_macro <- read_csv(file = "/home/yincy/git/R-codes/sslab/macrophage/20191217_M_T_frame_tagcount.csv") %>% 
    select(geneid:M_1029_T2_Positive)
```

```{r preprocessing macrophage data}
# aggregate reads count of duplicated genes
reads_table_macro_aggre <- aggregate(reads_table_macro[, 2:5], by = reads_table_macro[, 1], FUN = sum)
colnames(reads_table_macro) <- c("geneid", rep(c("T1_negative", "T2_negative"), each = 2))
```

Filtering to remove low expressed genes  
**As a general rule, a good threshold can be chosen by identifying the CPM that corresponds to a count of 10**.  
In this case cpm value below 1.7 will consider to be not expressed.   
```{r define expressed genes in macrophage}
cs <- colSums(reads_table_macro_aggre[, 2:5])
cpm_macro <- cpm(reads_table_macro_aggre[, 2:5])
keep <- rowSums(cpm_macro[, 1:2] > 1.7) >= 2 | rowSums(cpm_macro[, 3:4] > 1.7) >= 2
table(keep)
expressed_genes_macro <- reads_table_macro_aggre[keep, ] %>% .[, 1, drop = T]
```



# ligands in PT and receptor in macrophage  
## define genes of interest and background genes  
```{r define background genes}
background_genes <- expressed_genes_macro[expressed_genes_macro %in% rownames(ligand_target_matrix)]
```


#### define interested genes as differentially expressed genes between T-neg and T-pos  
```{r load data into DGEList}
group <- rep(c("T_neg", "T_pos"), each = 2)

reads_table_macro_aggre <- reads_table_macro_aggre %>% 
    as.data.frame() %>% 
    magrittr::set_rownames(value = reads_table_macro_aggre[, 1, drop = T]) %>% 
    select(-geneid) %>% 
    as.matrix()

y <- DGEList(counts = reads_table_macro_aggre[keep, ], 
             group = group)

rownames(y$samples) <- c("T1_negative", "T2_negative", "T1_positive", "T2_positive")
y$samples
```


Quality Control  
```{r barplot of library size}
par(mar = c(6, 4, 4, 2))
barplot(y$samples$lib.size, names.arg = rownames(y$samples), 
        las = 2, 
        main = "Bar plot of library size")
```


```{r dendragrame}
macro_dist <- dist(x = t(cpm(reads_table_macro_aggre[keep, ])))
macro_hclust <- hclust(d = macro_dist, method = "ward.D2")
fviz_dend(x = macro_hclust, 
          horiz = T, 
          cex = 0.71, 
          k = 2)
```

```{r heatmap of macrophage}
pheatmap::pheatmap(mat = cpm(reads_table_macro_aggre[keep, ]), 
                   show_rownames = F, 
                   scale = "row", 
                   clustering_distance_cols = "euclidean", 
                   clustering_method = "ward.D2")
```


```{r Normalize for composition bias}
y <- calcNormFactors(object = y)
y$samples
```


```{r differential expression analysis}
design <- model.matrix(~ 0 + group)
colnames(design) <- gsub("group", "", colnames(design))
design

v <- voom(counts = y, design = design, plot = T)
fit <- lmFit(object = v)
names(fit)

# make contrast
cont_matrix <- makeContrasts(neg_pos = T_neg - T_pos, levels = design)

fit_cont <- contrasts.fit(fit = fit, contrasts = cont_matrix)
fit_cont <- eBayes(fit = fit_cont)
dim(fit_cont)

# define differentially expressed if p < 0.1
summ.fit <- decideTests(object = fit_cont, p.value = 0.1)
summary(summ.fit)
```

```{r get up and down regulated genes}
genes_up_neg <- slot(summ.fit, ".Data") %>% as.data.frame() %>% 
    mutate(geneid = rownames(.)) %>% 
    filter(neg_pos == 1) %>% 
    pull(geneid)

genes_down_neg <- slot(summ.fit, ".Data") %>% 
    as.data.frame() %>% 
    mutate(geneid = rownames(.)) %>% 
    filter(neg_pos == -1) %>% 
    pull(geneid)
```

```{r}
diff_gene_expression <- reads_table_macro_aggre %>% 
    as.data.frame() %>% 
    mutate(geneid = rownames(.)) %>% 
    filter(geneid %in% c(genes_up_neg, genes_down_neg)) %>% 
    magrittr::set_rownames(value = .[, 5]) %>% 
    select(-geneid)
```

```{r}
par(mar = c(10, 0, 0, 0))
diff_gene_expression %>% 
    apply(1, function(x){(x / cs)*10^6}) %>% 
    t() %>%  
    pheatmap::pheatmap(breaks = seq(0, 290, by = 10), 
                       color = rev(hcl.colors(n = 30)),
                       border_color = NA,
                       show_rownames = F, 
                       angle_col = 45,
                       main = "difference expresed geens between PT and DCT")
```


```{r}
gene_oi <- c(genes_up_neg, genes_down_neg) %>% unique()
```


## define a set of potential ligands  
```{r}
lr_network <- readRDS(file = "/home/yincy/git/Data/NicheNet/data/lr_network.rds")
lr_network$from <- convert_human_to_mouse_symbols(lr_network$from)
lr_network$to <- convert_human_to_mouse_symbols(lr_network$to)
lr_network <- lr_network %>% 
    filter(!(is.na(from) | is.na(to)))

ligands <- lr_network %>% pull(from) %>% unique()
expressed_ligands <- intersect(x = toupper(ligands), y = toupper(expressed_genes_pt))

lr_network_expressed <- lr_network %>% 
    filter(toupper(from) %in% toupper(expressed_genes_pt) & toupper(to) %in% toupper(expressed_genes_macro))
```


```{r}
potential_ligand <- lr_network_expressed %>% 
    pull(from) %>% 
    unique()
```


## Perform NicheNet's ligand activity analysis on the gene set of interest   
```{r}
ligand_activity <- predict_ligand_activities(geneset = gene_oi, 
                                             background_expressed_genes = background_genes, 
                                             ligand_target_matrix = ligand_target_matrix, 
                                             potential_ligands = potential_ligand)
ligand_activity %>% 
    arrange(desc(pearson))
```

```{r}
ligand_activity %>% 
    pull(pearson) %>% 
    hist(breaks = 30, col = "orange", main = "Histogram of Pearson")
abline(v = ligand_activity %>% arrange(desc(pearson))%>% .[21, ] %>% pull(pearson), lty = 2,lwd = 2, col = "red")
```


```{r}
best_upstream_ligands <- ligand_activity %>% 
    arrange(desc(pearson)) %>%
    pull(test_ligand) %>% 
    head(n = 21)
```


## infer target geens of top ligands  
```{r}
active_ligands_target_df <- best_upstream_ligands %>% 
    lapply(FUN = get_weighted_ligand_target_links, 
           geneset = gene_oi, 
           ligand_target_matrix = ligand_target_matrix, 
           n = 250) %>% 
    bind_rows()

active_ligands_target_df %>% 
    arrange(desc(weight)) %>% 
    filter(ligand == "Hmgb2")
```

```{r}
write.csv(x = active_ligands_target_df %>% arrange(-weight), 
          file = "/home/yincy/git/R-codes/sslab/macrophage/active_ligands_targte_df.csv", 
          row.names = F, quote = F)
```


```{r}
tiff(filename = "/home/yincy/git/R-codes/sslab/macrophage/active_ligands_targets_heatmap.tiff")
active_ligands_target_df %>% 
    ggplot(aes(reorder(ligand, -weight), reorder(target, -weight))) +
    geom_tile(aes(fill = weight)) +
    scale_fill_gradient(low = "blue", high = "red") +
    scale_x_discrete(name = "ligands", expand = c(0, 0)) +
    scale_y_discrete(name = "target", expand = c(0, 0)) +
    theme(axis.text.x = element_text(angle = -30, hjust = 0), 
          panel.background = element_rect(fill = "white"), 
          panel.grid.major = element_line(color = "grey80", linetype = 2, size= 0.1), 
          axis.ticks.length = unit(0, "cm"))
dev.off()
```

## infer liagnds signaling path  
```{r}
ligands_all <- c("Hmgb2")
targets_all <- c("Mki67", "Top2a", "Iqgap3", "Zfhx3", "Dlc1")

ligand_tf_matrix <- readRDS(file = "/home/yincy/git/Data/NicheNet/data/ligand_tf_matrix.rds")
rownames(ligand_tf_matrix) <- convert_human_to_mouse_symbols(rownames(ligand_tf_matrix))
colnames(ligand_tf_matrix) <- convert_human_to_mouse_symbols(colnames(ligand_tf_matrix))
ligand_tf_matrix <- ligand_tf_matrix[!is.na(rownames(ligand_tf_matrix)), !is.na(colnames(ligand_tf_matrix))]

sig_network <- readRDS(file = "/home/yincy/git/Data/NicheNet/data/signaling_network.rds")
sig_network$from <- convert_human_to_mouse_symbols(sig_network$from)
sig_network$to <- convert_human_to_mouse_symbols(sig_network$to)
sig_network <- sig_network %>% 
    filter(!(is.na(from) | is.na(to)))

gr_network <- readRDS(file = "/home/yincy/git/Data/NicheNet/data/gr_network.rds")
gr_network$from <- convert_human_to_mouse_symbols(gr_network$from)
gr_network$to <- convert_human_to_mouse_symbols(gr_network$to)
gr_network <- gr_network %>% 
    filter(!(is.na(from) | is.na(to)))

weighted_networks <- construct_weighted_networks(
    lr_network = lr_network, 
    gr_network = gr_network, 
    sig_network = sig_network,
    source_weights_df = source_weights_df
)

weighted_networks <- apply_hub_corrections(weighted_networks = weighted_networks, 
                                           lr_sig_hub = hyperparameter_list$lr_sig_hub, 
                                           gr_hub = hyperparameter_list$gr_hub)

active_liagnds_signaling_network <- get_ligand_signaling_path(
    ligand_tf_matrix = ligand_tf_matrix, 
    ligands_all = ligands_all, 
    targets_all = targets_all, 
    weighted_networks = weighted_networks
)
```


```{r visualization signal network}
df <- bind_rows(active_liagnds_signaling_network$sig, active_liagnds_signaling_network$gr)
df <- aggregate(df[, 3], by = df[, 1:2], FUN = sum)

sig_path <- graph_from_data_frame(d = df, directed = T)

sig_path <- simplify(sig_path, remove.multiple = T, remove.loops = T, edge.attr.comb = c(weight = sum))


receptors <- lr_network %>% 
    filter(from %in% (active_ligands_target_df %>% pull(ligand) %>% unique())) %>% 
    pull(to)

vcol <- rep("grey40", vcount(sig_path))
vcol[V(sig_path)$name %in% ligands_all] <- "orange"
vcol[V(sig_path)$name %in% targets_all] <- "green"
vcol[V(sig_path)$name %in% receptors] <- "blue"

ecol <- rep("grey70", ecount(sig_path))


l <- layout_with_lgl(sig_path)
l <- norm_coords(l)
```

```{r}
pdf(file = "/home/yincy/git/R-codes/sslab/macrophage/Hmgb2_signaling_path.pdf")
par(mar = c(0, 0, 0, 0))
plot(sig_path, 
     vertex.color = vcol, 
     vertex.frame.color = "grey", 
     vertex.size = 60, 
     rescale = F, 
     layout = l * 9,
     xlim = c(-10, 10), 
     ylim = c(-10, 10), 
     edge.arrow.size = 0.5, 
     edge.width = E(sig_path)$weight * 3, 
     edge.color = ecol)
dev.off()
```


## infer supporting database network  












